securebanking-openbanking-uk-gateway-conversion-filter/src/main/java/com/forgerock/securebanking/uk/gateway/conversion/filter/IntentConverterFilter.java

```diff
@@ -0,0 +1,239 @@
+ * {@code {
+ *      "name": "IntentConverterFilter-name"
+ *      "type": "IntentConverterFilter",
+ *      "config": {
+ *         "messageType: "request"
+ *      }
+ *  }
+ *  }
+ * </pre>
+ * <pre>
+ *  {@code {
+ *       "name": "IntentConverterFilter-name"
+ *       "type": "IntentConverterFilter",
+ *       "config": {
+ *          "messageType: "response"
+ *       }
+ *   }
+ *   }
+ *  </pre>
+ */
+public class IntentConverterFilter implements Filter {
+
+    private static final Logger logger = LoggerFactory.getLogger(IntentConverterFilter.class);
+
+    private final MessageType messageType;
+    private static final ObjectMapper MAPPER = GenericConverterMapper.getMapper();
+
+    /**
+     * Constructor
+     * @param messageType
+     */
+    public IntentConverterFilter(final MessageType messageType) {
+        this.messageType = messageType;
+    }
+
+    @Override
+    public Promise<Response, NeverThrowsException> filter(Context context, Request request, Handler next) {
+        if (logger.isInfoEnabled()) {
+            printFilterInfo();
+        }
+
+        if (messageType.equals(MessageType.REQUEST)) {
+            return process(request, request.getUri().asURI(), context, next);
+        }
+
+        // messageType RESPONSE
+        return next.handle(context, request)
+                .thenAsync(response -> process(response, request.getUri().asURI(), context, next));
+    }
+
+    /**
+     * Process the message
+     * @param message a {@link MessageImpl} to be processed
+     * @param uri the request URI to extract the api version to instance the proper converter
+     * @return a {@link Promise} to be passed the next handle
+     */
+    private Promise<Response, NeverThrowsException> process(final MessageImpl message, URI uri, Context context, Handler next) {
+        try {
+            String jsonPayload = getEntity(message);
+            if (Strings.isNullOrEmpty(jsonPayload)) {
+                throw new Exception("The entity to be converted should not be null");
+            }
+            OBVersion obVersion = ApiVersionUtils.getOBVersion(uri);
+            Object objectMapped = convert(getIntentType(jsonPayload), jsonPayload, obVersion);
+            logger.info("Result object {}", objectMapped.getClass().getSimpleName());
+            logger.debug("objectMapped {}", objectMapped);
+            // set the converted object in message (MessageType.REQUEST or MessageType.RESPONSE)
+            message.setEntity(toBytes(objectMapped));
+            if (!message.getHeaders().containsKey(ContentTypeHeader.NAME)) {
+                message.getHeaders().put(ContentTypeHeader.NAME, APPLICATION_JSON_CHARSET_UTF_8);
+            }
+            // MessageType.REQUEST
+            if (messageType.equals(MessageType.REQUEST)) {
+                return next.handle(context, (Request) message);
+            }
+            // MessageType.RESPONSE
+            return newResultPromise(((Response) message).setStatus(Status.OK));
```

Comment:
What happens if a Filter after this one in the chain or the RS responds with an error such as HTTP 400?

What if an OBErrorResponse1 json payload is the message.entity? Will this be allowed to pass through or will this cause an error in the conversion logic?

Reply:
Right! nice one! fixed!
- The filter will process the response only when the response from the called handle is successful otherwise pass through the response to the next handle.

Reply:
Great, I would remove calling setStatus just to be safe

Reply:
Set status removed.